### ISSUE TITLE
PERF: Taxonium always uses ~8% CPU even if it's in the background

### ISSUE THREAD

**Issue by corneliusroemer** [theosanderson/taxonium#432]
I was surprised to see Taxonium use 8% of CPU in the background even when the window is not showing.

Is this because the server is constantly checking for requests? Would be good to somehow reduce CPU usage.

Having 10% in the background can quickly add up to quite some energy consumption

**Comment by corneliusroemer** [comment]
It had used 30min of 100% CPU - more than all the other processes during a whole day of usage.

I used taxonium for maybe 30min in total, so the background adds up to quite a lot.

**Comment by theosanderson** [comment]
This definitely isn't expected behaviour. I can't (immediately) reproduce it in the sense that if I just load a tree and do some basic stuff then CPU usage drops to 0% when the tree is not being used. What I suspect is happening in your case is that after a while of interaction some operation has triggered a situation in which there is an infinite loop of the client requesting a search, or nodes, over and over again. Definitely would be great to fix that, but unfortunately they can be hard bugs to run down. If you find this CPU usage even before interacting with the tree then my explanation wouldn't make sense and that would be especially useful to know.

**Comment by corneliusroemer** [comment]
Definitely possible that it's related to search.

What's the best way to get logs? Could you maybe export logs to some file, so I can send them over once there's a crash of some sort?

**Comment by corneliusroemer** [comment]
So it's actually 15% usage all the time.

These are the processes per htop:
<img width="1727" alt="image" src="https://user-images.githubusercontent.com/25161793/196033209-103e3cfe-30e0-489b-b1a3-a85cdaad090b.png">

Some logging should help figure this out, no?

**Comment by corneliusroemer** [comment]
This happens even without search! Just opening taxonium and loading a tree.

Since this seems to be mostly the renderer, could it be that you constantly rerender despite there being no view changes at all?

<img width="1721" alt="image" src="https://user-images.githubusercontent.com/25161793/196036119-09b24e26-2685-41c4-8801-facda16c7672.png">

**Comment by corneliusroemer** [comment]
Yeah it has nothing to do with search, since the process for searching is a different one. Almost definitely something in render.

**Comment by theosanderson** [comment]
Oh weird - thanks. In my check I had loaded in the browser, which does not (in my experience) do this. But I can see something similar when I use the Electron client. Even though it is literally just loading the same page in a Chromium instance. Will inviestigate. The workaround is to open in the browser.

**Comment by corneliusroemer** [comment]
Not really a workaround as it's much slower and crashes :D Getting good logging into a file would be amazing!

**Comment by theosanderson** [comment]
Suggested workaround (not expressed clearly) was to open the `https://taxonium.org/?backend=http://localhost:xxxx` URL in the browser. That shouldn't be any slower as it's doing the same thing. However I'm no longer convinced that it doesn't use CPU usage, it may just be that it's split across a couple of processes at ~3% that I hadn't noticed.

Still looking into where that comes from. 

For writing _backend_ logs to a file you should be able to use:

`/Applications/Taxonium.app/Contents/MacOS/Taxonium ~/log.txt 2>&1`

**Comment by theosanderson** [comment]
For writing _backend_ logs to a file you should be able to use:

`/Applications/Taxonium.app/Contents/MacOS/Taxonium &> ~/log.txt`

**Comment by theosanderson** [comment]
Looked into this a bit.

- As above, backend seems fine mostly, this issue is about front-end issue
- No React re-rendering is going on when Idle in the frontend
- The only thing that is happening is a built-in re-rendering from DeckGL, which I think is the same issue discussed in https://github.com/visgl/deck.gl/issues/6961 . Fixing that upstream is out of scope atm, but we can monitor that issue for any updates.

**Comment by theosanderson** [comment]
Reopening this in case we think of anything else..
